mclagsyncd/mclaglink.cpp

```diff
@@ -401,276 +519,1390 @@ void MclagLink::setFdbEntry(char *msg, int msg_len)
+    bool                      is_active_role;
+    bool                      valid_system_id = false;
+    string                    system_id_str;
+    char                      *cur;
+    size_t                    cur_len = 0;
+    mclag_sub_option_hdr_t    *op_hdr;
+    vector<FieldValueTuple>   fvVector;
+
+    while (cur_len < msg_len)
+    {
+        cur = msg + cur_len;
+        op_hdr = reinterpret_cast<mclag_sub_option_hdr_t *>(static_cast<void *>(cur));
+
+        switch(op_hdr->op_type)
+        {
+            case MCLAG_SUB_OPTION_TYPE_MCLAG_ID:
+                memcpy(&mlag_id, op_hdr->data, op_hdr->op_len);
+                break;
+
+            case MCLAG_SUB_OPTION_TYPE_ICCP_ROLE:
+                memcpy(&is_active_role, op_hdr->data, op_hdr->op_len);
+                fvVector.push_back(
+                        make_pair("role", is_active_role ? "active" : "standby"));
+                break;
+
+            case MCLAG_SUB_OPTION_TYPE_SYSTEM_ID:
+                valid_system_id = true;
+                system_id_str = MacAddress::to_string(op_hdr->data);
+                fvVector.push_back(make_pair("system_mac", system_id_str));
+                break;
+            default:
+                SWSS_LOG_ERROR("Invalid option type %u", op_hdr->op_type);
+                break;
+        }
+        cur_len += (MCLAG_SUB_OPTION_HDR_LEN + op_hdr->op_len);
+    }
+    if ((mlag_id > 0) && (fvVector.size() > 0))
+    {
+        /* Update MLAG table: key = mlag_id, value = role */
+        p_mclag_tbl->set(to_string(mlag_id), fvVector);
+        SWSS_LOG_NOTICE("Set mlag %d ICCP role to %s, system_id(%s)",
+                mlag_id, is_active_role ? "active" : "standby",
+                valid_system_id ? system_id_str.c_str() : "None");
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid parameter, mlag %d", mlag_id);
+    }
+}
+
+/* Set the system_mac field in the STATE_MCLAG_TABLE */
+void MclagLink::mclagsyncdSetSystemId(
+        char                      *msg,
+        size_t                    msg_len)
+{
+    int                       mlag_id = 0;
+    string                    system_id_str;
+    char                      *cur;
+    size_t                    cur_len = 0;
+    mclag_sub_option_hdr_t    *op_hdr;
+    vector<FieldValueTuple>   fvVector;
+
+    while (cur_len < msg_len)
+    {
+        cur = msg + cur_len;
+        op_hdr = reinterpret_cast<mclag_sub_option_hdr_t *>(static_cast<void *>(cur));
+
+        switch(op_hdr->op_type)
+        {
+            case MCLAG_SUB_OPTION_TYPE_MCLAG_ID:
+                memcpy(&mlag_id, op_hdr->data, op_hdr->op_len);
+                break;
+
+            case MCLAG_SUB_OPTION_TYPE_SYSTEM_ID:
+                system_id_str = MacAddress::to_string(op_hdr->data);
+                fvVector.push_back(make_pair("system_mac", system_id_str));
+                break;
+
+            default:
+                SWSS_LOG_ERROR("Invalid option type %u", op_hdr->op_type);
```

Comment:
Unexpected option perhjaps should skip the rest of the code processing and bail out instead?